Cosmo rename arg to io

nstarman opened the pull request:
Adding an option to remap symbols in input and output of a cosmology.
This should be helpful for #14701.

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address ...

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #<Issue Number>

mhvk commented on astropy/cosmology/io/table.py:
```diff
@@ -11,20 +11,20 @@
     table : `~astropy.table.Table`
         The object to parse into a |Cosmology|.
-    index : int, str, or None, optional
+    index : int or str or None, optional
```
Not sure about changes like these - unrelated to the PR and not an improvement since grammatically incorrect. Is this to help computers or humans?

nstarman replied:
This comes from https://numpydoc.readthedocs.io/en/latest/format.html#parameters, where the separation between types is an "or" and between the list of types and info like "optional" is a comma.

nstarman replied:
Also, yes, this consistency does help with programmatic parsing of docstrings.

nstarman replied:
Since the docstring type annotation mimics the real type annotation, I've also seen a lot of new docstrings swap the "or" for "|"
```
index : int | str | None, optional
```

nstarman replied:
I think this is actually the most readable.

mhvk replied:
I don't see the text state that more than 2 options need multiple "or".  Am I reading over it?

mhvk replied:
p.s. I sort-of like the `|` option, but feel it takes away a bit from the purpose of a docstring, which is to give a description in english, aimed at humans... Which is why I prefer "A, B or C".

nstarman replied:
> I don't see the text state that more than 2 options need multiple "or". Am I reading over it?

Hm. I guess it's not explicit. I glanced at Pandas and they use commas in their lists. IMO it slows down my parsing of types versus comments like optional, keyword-only, etc. with "or" or "|" I see it as a block. With "," they are mixed together.

```
param : this or that or something, optional
```

vs

```
param : this, that, or something, optional
```

The comma works as a delimiter, making the second option muddled compared to the first.

mhvk replied:
For me it is less obvious, although I would put "optional" in parentheses (as you had it) or, more likely, use a semi-colon.  But repeating "or" is not really grammatical... Note that it is slightly better if you do not use a comma before "or" (which is ApJ-style, but not that commonly done elsewhere).

nstarman replied:
I've reverted this.

mhvk commented on astropy/cosmology/io/table.py:
```diff
@@ -11,20 +11,20 @@
         is a string, the "name" column is used as the indexing column.
 
-    move_to_meta : bool (optional, keyword-only)
+    move_to_meta : bool, optional keyword-only
```
Same with this: original just seems better

nstarman replied:
This comes from https://numpydoc.readthedocs.io/en/latest/format.html#parameters, where the separation should be a comma.

mhvk replied:
OK, that's fine. Here I mostly disliked "optional keyword-only" without comma. Do we need `keyword-only`? That duplicates the signature. Anyway, this one I mind much less.

nstarman replied:
Hopefully codified in numpydoc eventually.
https://github.com/numpy/numpydoc/pull/367

nstarman replied:
The lack of comma... yeah, same complaint as above... All this mixing of types and comments is harder to chunk than pre-chunked info!

```
move_to_meta : <type block>, <comment block>
```

vs

```
move_to_meta : mixed, type, and, comment, block
```

I'm not opposed to changing it, but I at least find it less useful than pre-chunking the info.

nstarman replied:
I'll roll this back for now, and push a PR later that rectifies the docs.

mhvk commented on astropy/cosmology/io/table.py:
```diff
@@ -119,46 +126,53 @@ def from_table(table, index=None, *, move_to_meta=False, cosmology=None):
             index = indices[0]
         elif len(indices) == 0:  # matches from loc_indices
-            raise KeyError(f"No matches found for key {indices}")
+            msg = f"No matches found for key {indices}"
```
This is another change that is not related to the PR and, as far as I can see, just adds line and complexity (however slight) for no obvious benefit.

nstarman replied:
Agreed it's not so related to the PR, but it does add benefit.
Python includes the raise in the traceback. By using a string literal, the error message is duplicated in the traceback, which is less readable. I tripped this in a test and wanted to improve the user experience.
It would be better if Python had an option to not include the string literal in the traceback, but short of a change to Python, this is the way to make nicer tracebacks.

mhvk replied:
I'm not sure; what is wrong with the following traceback?
```
n [4]: u.Quantity('bla')
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File /usr/lib/python3/dist-packages/astropy/units/quantity.py:490, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    489 try:
--> 490     value = float(v.group())
    492 except Exception:

AttributeError: 'NoneType' object has no attribute 'group'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In [4], line 1
----> 1 u.Quantity('bla')

File /usr/lib/python3/dist-packages/astropy/units/quantity.py:493, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    490     value = float(v.group())
    492 except Exception:
--> 493     raise TypeError(
    494         f'Cannot parse "{value}" as a {cls.__name__}. It does not '
    495         "start with a number."
    496     )
    498 unit_string = v.string[v.end() :].strip()
    499 if unit_string:

TypeError: Cannot parse "bla" as a Quantity. It does not start with a number.

```

nstarman replied:
With the change this should look like 

```
>>> import astropy.units as u
>>> u.Quantity("bla")
Traceback (most recent call last):
  File "/Users/nmrs/local/astropy/astropy/units/quantity.py", line 479, in __new__
    value = float(v.group())
AttributeError: 'NoneType' object has no attribute 'group'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/nmrs/local/astropy/astropy/units/quantity.py", line 486, in __new__
    raise TypeError(msg)
TypeError: Cannot parse "bla" as a Quantity. It does not start with a number.
```,

getting rid of the duplication of the Exception message. We see the actual  `TypeError`, there's not much point showing it's generation during the traceback.

nstarman replied:
Python has been careful to not show extraneous information:

```
Traceback (most recent call last):
  File "/Users/nmrs/local/astropy/astropy/units/quantity.py", line 479, in __new__
    value = float(v.group())
AttributeError: 'NoneType' object has no attribute 'group'
```

doesn't show how AttributeError is produced. Unfortunately they use C stuff to get rid of even the `raise AttributeError(msg)` step and I don't think we can achieve this degree of concision.

mhvk replied:
Since the error message give context, there is actually very little difference. If I change `Quantity` to use `msg`, I get:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~/data/mhvk/packages/astropy/astropy/units/quantity.py:479, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    478 try:
--> 479     value = float(v.group())
    481 except Exception:

AttributeError: 'NoneType' object has no attribute 'group'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 u.Quantity('bla')

File ~/data/mhvk/packages/astropy/astropy/units/quantity.py:486, in Quantity.__new__(cls, value, unit, dtype, copy, order, subok, ndmin)
    481 except Exception:
    482     msg = (
    483         f'Cannot parse "{value}" as a {cls.__name__}. It does not '
    484         "start with a number."
    485     )
--> 486     raise TypeError(msg)
    488 unit_string = v.string[v.end() :].strip()
    489 if unit_string:

TypeError: Cannot parse "bla" as a Quantity. It does not start with a number.
```

nstarman replied:
Your view is different than mine! I made mine from the command line and it does not include the above context.

nstarman replied:
I'll roll this back for now. We can have discussion on https://beta.ruff.rs/docs/rules/#flake8-errmsg-em in a separate Issue. It has an auto fix, so if we decide to turn it on, it's very easily maintained.

nstarman replied:
I've reverted here.

mhvk commented on astropy/cosmology/io/mapping.py:
```diff
@@ -135,6 +162,8 @@ def to_mapping(
         `False`, default) or to merge with the rest of the mapping, preferring
         the original values (if `True`)
+    rename : dict or None, optional, keyword-only
+        A dictionary mapping keys in the Cosmology class' to keys.
```
Here, the description is a bit unclear: "keys of the map"? Generally, is "keys in the Cosmology class" the right term? Would "attributes" be better?

nstarman replied:
Agreed. Let me clean that up.

nstarman replied:
I changed it to "fields", which is the parlance of Python dataclasses.

mhvk replied:
That sounds better. I was wondering about `parameter` since this is about models; but mentioning this just in case you think it is better.

mhvk reviewed (commented):
@nstarman - Overall, I like the idea but found the large number of unrelated code and docstring changes distracting and ended up running out of time. So, mostly a few grumpy comments inline, plus a slightly larger question: don't we have something like this already in the general IO? I tried looking and couldn't find it, so probably not, but let me ping @taldcroft regardless, since perhaps this option to rename things would be something we should support more generally.

My mai

pllim commented:
> but found the large number of unrelated code and docstring changes distracting

I did a drive-by glance but gave up for the same reason too. It was unclear to me why these are necessary for the renaming process.

nstarman commented:
> Overall, I like the idea but found the large number of unrelated code and docstring changes distracting and ended up running out of time.

Apologies for the constellation of changes surrounding the main portion of the PR. These changes arose because when I added the docstring for the new argument I wanted to make sure I got the format correct, following numpydoc to the T. I found that the docstring was incorrectly implemented 👎  and rather than mixing styles within one docstring, I fixed the relevant docstrings.

eerovaher commented:
@mhvk wrote:
> ...don't we have something like this already in the general IO? I tried looking and couldn't find it, so probably not, but let me ping @taldcroft regardless, since perhaps this option to rename things would be something we should support more generally.

I also think renaming things at the `astropy.io` level could be useful. For example, I would find it convenient if I could work in Python with a `Table` with columns named `"ra"` and `"dec"`, but when I write a LaTeX file using `astropy.io` I might want the column names to appear as `"$\alpha$"` and `"$\delta$"`. So if there is a way of implementing renaming things so that both `astropy.table` and `astropy.cosmology` would benefit then that should be considered. I am not familiar enough with `astropy.io` to know if this would be simple to achieve or not, so it would be good to know what @taldcroft thinks.

nstarman commented:
@eerovaher, I agree, but m any of the Cosmology I/O methods do not rely on Table, so lacking prior art from table, I think this PR is a good place to start.

eerovaher commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,39 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
+        with pytest.raises(ValueError, match="disjoint"):
+            to_format("mapping", rename=to_rename)
+
+    def test_from_mapping_rename_disjoint(self, cosmo, to_format, from_format):
+        """Test roundtrip of ``rename`` in ``to/from_mapping()``."""
+        # rename
+        m = to_format("mapping")
+
+        with pytest.raises(ValueError, match="disjoint"):
+            from_format(m, format="mapping", rename={"name": "name"})
```
Could this and the preceding test be combined to a single parametrized test?

nstarman replied:
Maybe, in what way do you suggest?

eerovaher commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,39 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
+        assert "name" not in m.keys()
+        assert "cosmo_name" in m.keys()
+
+        # Wrong names = error
+        with pytest.raises(TypeError, match="there are unused parameters"):
+            from_format(m, format="mapping")
+
+        # Roundtrip. correct names = success
+        from_rename = {v: k for k, v in to_rename.items()}
+        got = from_format(m, format="mapping", rename=from_rename)
+        assert got == cosmo
```
I'd prefer if these were in separate tests. The problem with tests that assert too many things is that if an early assert fails then the following asserts are not executed at all.

nstarman replied:
The steps are tested in the prior tests, so except in a parallel environment or one that randomizes the test order, each piece is already tested. This is a roundtrip test but with `assert` along the way so that if it fails it does so close to the actual problem.

nstarman replied:
SGTM.

nstarman replied:
Done!

eerovaher commented on astropy/cosmology/io/ecsv.py:
```diff
@@ -9,8 +9,10 @@
-def read_ecsv(filename, index=None, *, move_to_meta=False, cosmology=None, **kwargs):
-    """Read a `~astropy.cosmology.Cosmology` from an ECSV file.
+def read_ecsv(
+    filename, index=None, *, move_to_meta=False, cosmology=None, rename=None, **kwargs
```
I didn't spot a single line of code that would modify `rename`, so the following should be safe and would allow some of the code to be slightly simplified:
```suggestion
    filename, index=None, *, move_to_meta=False, cosmology=None, rename={}, **kwargs
```

nstarman replied:
I would like for this to be a safer Python feature, and I agree it is in this case, but I never use a mutable value as a default value. IMO it's too much an anti-pattern.

eerovaher replied:
If `rename` were only used in a single function then I might insist on using `{}` as the default value, but it gets passed around between quite a few functions, so using `None` as the default value is not a bad idea.

eerovaher reviewed (commented):
Overall I think this is a good idea, but I do have some suggestions regarding the implementation.

nstarman commented:
Thanks @eerovaher for the detailed review!

nstarman commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,39 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
         assert got.meta["mismatching"] == "will error"
 
+    def test_to_mapping_rename_disjoint(self, cosmo, to_format):
+        """Test roundtrip of ``rename`` in ``to/from_mapping()``."""
```
Agreed. Changing the docstrings.

nstarman commented on astropy/cosmology/io/row.py:
```diff
@@ -59,8 +62,13 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
+
     # special values
-    name = row["name"] if "name" in row.columns else None  # get name from column
+    name = row[kn] if kn in row.columns else None  # get name from column
```
Turns out that's not the same. Reverting.

eerovaher commented on astropy/cosmology/io/mapping.py:
```diff
@@ -79,18 +105,12 @@ class itself.
-    params = dict(map)  # so we are guaranteed to have a poppable map
+    # Rename keys, if given a ``renames`` dict.
+    # Also, make a copy of the mapping, so we can pop from it.
+    params = _rename_map(dict(map), renames={} if rename is None else rename)
```
A bit shorter:
```suggestion
    params = _rename_map(dict(map), renames=rename or {})
```

eerovaher commented on astropy/cosmology/io/mapping.py:
```diff
@@ -216,7 +253,9 @@ def to_mapping(
         m["meta"] = meta  # TODO? should meta be type(cls)
-
+    # Rename keys
+    if rename is not None:
+        m = _rename_map(m, rename)
     return m
```
Equivalent outcome, but fewer lines of code:
```suggestion
    return m if rename is None else _rename_map(m, rename)
```

eerovaher commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,36 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
+    def test_to_mapping_rename_conflict(self, cosmo, to_format):
+        """Test ``rename`` in ``to_mapping()``."""
+        to_rename = {"name": "name"}
+        with pytest.raises(ValueError, match="disjoint"):
```
I would prefer if the `match` argument (here and also in other tests) contained a large enough fragment of the error message so that even someone not familiar with the code or the test suite would have a good idea what exactly is this test checking. In my experience a very detailed `match` argument helps a lot in making the test suite self-documenting.

nstarman replied:
SGTM.

nstarman replied:
Done=!

nstarman commented on astropy/cosmology/io/row.py:
```diff
@@ -59,8 +62,13 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
+    inv_rename = {v: k for k, v in rename.items()} if rename is not None else {}
+    kn = inv_rename.get("name", "name")
+    km = inv_rename.get("meta", "meta")
+    kc = inv_rename.get("cosmology", "cosmology")
```
Changed.

eerovaher commented on astropy/cosmology/io/mapping.py:
```diff
@@ -42,6 +65,9 @@ class itself.
         filling in any non-mandatory arguments missing in 'map'.
 
+    rename : dict or None (optional, keyword-only)
+        A dictionary mapping keys in 'map' to fields of the Cosmology.
```
```suggestion
        A dictionary mapping keys in ``map`` to fields of the
        `~astropy.cosmology.Cosmology`.
```

eerovaher commented on astropy/cosmology/io/mapping.py:
```diff
@@ -17,7 +17,30 @@
 
-def from_mapping(map, *, move_to_meta=False, cosmology=None):
+def _rename_map(map, /, renames):
+    # So we are guaranteed to have a poppable map.
```
What's this comment about?

nstarman replied:
That the output guarantees a dict that is decoupled from the input map.

eerovaher commented on astropy/cosmology/io/row.py:
```diff
@@ -10,8 +10,8 @@
-def from_row(row, *, move_to_meta=False, cosmology=None):
-    """Instantiate a `~astropy.cosmology.Cosmology` from a `~astropy.table.Row`.
+def from_row(row, *, move_to_meta=False, cosmology=None, rename=None):
+    """Instantiate a |Cosmology| from a `~astropy.table.Row`.
```
```suggestion
    """Instantiate a `~astropy.cosmology.Cosmology` from a `~astropy.table.Row`.
```

eerovaher commented on astropy/cosmology/io/row.py:
```diff
@@ -26,14 +26,17 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
         filling in any non-mandatory arguments missing in 'table'.
 
+    rename : dict or None (optional, keyword-only)
+        A dictionary mapping columns in the row to fields of the Cosmology.
```
```suggestion
        A dictionary mapping columns in the row to fields of the `~astropy.cosmology.Cosmology`.
```

eerovaher commented on astropy/cosmology/io/row.py:
```diff
@@ -26,14 +26,17 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
     Returns
     -------
-    `~astropy.cosmology.Cosmology` subclass instance
+    |Cosmology| subclass instance
```
```suggestion
    `~astropy.cosmology.Cosmology`
```

eerovaher commented on astropy/cosmology/io/row.py:
```diff
@@ -59,8 +62,13 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
+
     # special values
-    name = row["name"] if "name" in row.columns else None  # get name from column
+    name = row[kname] if kname in row.columns else None  # get name from column
```
https://github.com/astropy/astropy/commit/d87cde4144b926eb2ca1831bbb18483cd9ad2b2b is merged now, so after rebasing it would be possible to write
```suggestion
    name = row.get(kname)
```

eerovaher commented on astropy/cosmology/io/row.py:
```diff
@@ -72,20 +80,23 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
     Parameters
     ----------
-    cosmology : `~astropy.cosmology.Cosmology` subclass instance
+    cosmology : |Cosmology|
```
```suggestion
    cosmology : `~astropy.cosmology.Cosmology`
```

eerovaher commented on astropy/cosmology/io/row.py:
```diff
@@ -99,14 +110,14 @@ def to_row(cosmology, *args, cosmology_in_meta=False, table_cls=QTable):
     --------
-    A Cosmology as a `~astropy.table.Row` will have the cosmology's name and
-    parameters as columns.
+    A Cosmology as a `~astropy.table.Row` will have the cosmology's name and parameters
```
```suggestion
    A `~astropy.cosmology.Cosmology` as a `~astropy.table.Row` will have the cosmology's name and parameters
```

eerovaher commented on astropy/cosmology/io/table.py:
```diff
@@ -33,14 +33,17 @@ def from_table(table, index=None, *, move_to_meta=False, cosmology=None):
         will be ``{'key': 10}``).
 
-    cosmology : str, `~astropy.cosmology.Cosmology` class, or None (optional, keyword-only)
+    cosmology : str or |Cosmology| class or None (optional, keyword-only)
```
I don't think the `type` needs to be specified more thoroughly than `str` in this line.
```suggestion
    cosmology : str or type or None (optional, keyword-only)
```

eerovaher commented on astropy/cosmology/io/table.py:
```diff
@@ -33,14 +33,17 @@ def from_table(table, index=None, *, move_to_meta=False, cosmology=None):
         filling in any non-mandatory arguments missing in 'table'.
 
+    rename : dict or None (optional, keyword-only)
+        A dictionary mapping columns in 'table' to fields of the Cosmology class.
```
```suggestion
        A dictionary mapping columns in 'table' to fields of the `~astropy.cosmology.Cosmology` class.
```

eerovaher commented on astropy/cosmology/io/table.py:
```diff
@@ -33,14 +33,17 @@ def from_table(table, index=None, *, move_to_meta=False, cosmology=None):
     Returns
     -------
-    `~astropy.cosmology.Cosmology` subclass instance
+    |Cosmology| subclass instance
```
```suggestion
    `~astropy.cosmology.Cosmology`
```

eerovaher commented on astropy/cosmology/io/table.py:
```diff
@@ -110,7 +113,11 @@ def from_table(table, index=None, *, move_to_meta=False, cosmology=None):
-            indices = np.where(table["name"] == index)[0]
+            nc = "name"  # default name column
+            if rename is not None:  # from inverted `rename`
+                nc = {v: k for k, v in rename.items()}.get(nc, nc)
```
I don't quite like constructing a temporary `dict` here.
```suggestion
                for key, value in rename.items():
                    if value == "name":
                        nc = key
                        break
```

eerovaher commented on astropy/cosmology/io/table.py:
```diff
@@ -232,7 +240,15 @@ def to_table(cosmology, *args, cls=QTable, cosmology_in_meta=True):
-    tbl.add_index("name", unique=True)
+
+    # Renames
+    renames = {} if rename is None else rename
```
Shorter:
```suggestion
    renames = rename or {}
```

eerovaher commented on astropy/cosmology/io/tests/test_ecsv.py:
```diff
@@ -127,6 +127,29 @@ def test_readwrite_ecsv_instance(
         got = read(fp)
         assert got == cosmo
 
+    def test_readwrite_ecsv_rename(
```
```suggestion
    def test_readwrite_ecsv_renamed_columns(
```

eerovaher commented on astropy/cosmology/io/tests/test_ecsv.py:
```diff
@@ -127,6 +127,29 @@ def test_readwrite_ecsv_instance(
+        assert "cosmo_name" in tbl.colnames
+
+        # Errors if reading
+        with pytest.raises(TypeError, match="there are unused parameters"):
```
I'd like the `match` value to be even more detailed. Looking at this test it is not at all clear to me where the unused parameters are.

nstarman replied:
Done!

eerovaher commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,36 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
 
+    def test_to_mapping_rename_conflict(self, cosmo, to_format):
+        """Test ``rename`` in ``to_mapping()``."""
+        to_rename = {"name": "name"}
```
It would be good to add an entry here that does not cause a conflict so that the error message below could be tested more thoroughly.

nstarman replied:
Done!

eerovaher commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,36 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
+    def test_to_mapping_rename_conflict(self, cosmo, to_format):
+        """Test ``rename`` in ``to_mapping()``."""
+        to_rename = {"name": "name"}
+        with pytest.raises(ValueError, match="'renames' values must be disjoint"):
```
Here the `match` value is clear enough to convey the purpose of the test, but we also want the error message to tell the user what the conflicting names are, so we should test that by expanding the `match` value even more.
```suggestion
        with pytest.raises(ValueError, match="'renames' values must be disjoint"):
```

nstarman replied:
Done!

eerovaher commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,36 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
+        with pytest.raises(ValueError, match="'renames' values must be disjoint"):
+            to_format("mapping", rename=to_rename)
+
+    def test_from_mapping_rename_disjoint(self, cosmo, to_format, from_format):
```
```suggestion
    def test_from_mapping_rename_conflict(self, cosmo, to_format, from_format):
```

eerovaher reviewed (changes requested):
I still have quite a few suggestions, but most of them are just changes to how to write hyperlinks in the docstrings.

I also recommend using even larger `match` values with `pytest.raises()`. Too short error or warning message fragments have caused problems before. For example, the `match` values added in d0de29e9066e7f8e5c7d3c50c7582e67fbfb0298 were too short, so the fact that the warning messages were malformed wasn't noticed until almost a year later: #14662.

nstarman commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,36 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
+        """Test ``rename`` in `from_mapping()``."""
+        m = to_format("mapping")
+
+        with pytest.raises(ValueError, match="'renames' values must be disjoint"):
```
Done!

eerovaher commented on astropy/cosmology/io/tests/test_table.py:
```diff
@@ -206,6 +206,23 @@ def test_tofrom_table_mutlirow(self, cosmo_cls, cosmo, from_format, add_index):
         with pytest.raises(ValueError, match="more than one"):
             from_format(tbls, index=cosmo.name, format="astropy.table")
 
+    def test_tofrom_table_rename(self, cosmo, to_format, from_format):
```
This test is very similar to `test_tofrom_row_rename()`, so it might be a good idea to write a single parametrized test, but that wouldn't work because the tests are in different files. Maybe it would be a good idea to organize the test differently, but even if it is then that would be beyond the scope of this pull request.

eerovaher reviewed (commented):
This is getting close to being finished, I just have a few comments and suggestions left.

nstarman commented on astropy/cosmology/io/row.py:
```diff
@@ -59,8 +62,13 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
     """
+    inv_rename = {v: k for k, v in rename.items()} if rename is not None else {}
+    kname = inv_rename.get("name", "name")
+    kmeta = inv_rename.get("meta", "meta")
```
I do think it ends up looking cleaner.

nstarman commented on astropy/cosmology/io/row.py:
```diff
@@ -118,12 +129,14 @@ def to_row(cosmology, *args, cosmology_in_meta=False, table_cls=QTable):
-    The cosmological class and other metadata, e.g. a paper reference, are in
-    the Table's metadata.
+    The cosmological class and other metadata, e.g. a paper reference, are in the
+    Table's metadata.
```
✅

nstarman commented on astropy/cosmology/io/row.py:
```diff
@@ -99,14 +110,14 @@ def to_row(cosmology, *args, cosmology_in_meta=False, table_cls=QTable):
     --------
-    A Cosmology as a `~astropy.table.Row` will have the cosmology's name and
-    parameters as columns.
+    A `~astropy.cosmology.Cosmology` as a `~astropy.table.Row` will have the cosmology's name and parameters
```
✅

nstarman commented on astropy/cosmology/io/tests/test_mapping.py:
```diff
@@ -129,6 +129,40 @@ def test_tofrom_mapping_move_tofrom_meta(self, cosmo, to_format, from_format):
+
+        match = "'renames' values must be disjoint from 'map' keys, the common keys are: {'name'}"
+        with pytest.raises(ValueError, match=match):
+            from_format(m, format="mapping", rename={"name": "name"})
```
✅

nstarman commented on astropy/cosmology/io/ecsv.py:
```diff
@@ -9,8 +9,10 @@
+def read_ecsv(
+    filename, index=None, *, move_to_meta=False, cosmology=None, rename=None, **kwargs
+):
+    """Read a |Cosmology| from an ECSV file.
```
✅

eerovaher commented on astropy/cosmology/io/table.py:
```diff
@@ -110,7 +113,14 @@ def from_table(table, index=None, *, move_to_meta=False, cosmology=None):
-            indices = np.where(table["name"] == index)[0]
+            nc = "name"  # default name column
+            if rename is not None:  # from inverted `rename`
+                for key, value in rename.items():
+                    if value == "name":
+                        nc = key
+                        break
```
Looks like this code is untested.

nstarman replied:
Now tested!

eerovaher reviewed (commented):
This is almost ready for merging, I only have three comments left. 

These is an untested code branch that we haven't apparently noticed until now.

It looks to me like some of the lines involving strings (docstrings or `match` values for `pytest.raises()`) are too long. Black doesn't complain because it avoids splitting strings and I am not aware of a convenient way of displaying line lengths in the GitHub web interface, but any decent IDE should recognize if this is indeed the case.

I would try to get rid of the `Apply suggestions from code review` commit (not just here but in every pull request), but rewriting the Git repository history might not be worth the effort.

nstarman commented:
I squashed commits, added and modified tests, added to the Examples in the docstring, and put in a "whatsnew".

eerovaher commented on astropy/cosmology/io/row.py:
```diff
@@ -26,14 +26,17 @@ def from_row(row, *, move_to_meta=False, cosmology=None):
         will be ``{'key': 10}``).
 
-    cosmology : str, `~astropy.cosmology.Cosmology` class, or None (optional, keyword-only)
+    cosmology : str, |Cosmology| class, or None (optional, keyword-only)
```
Shouldn't this be written as
```suggestion
    cosmology : str, type, or None (optional, keyword-only)
```

eerovaher reviewed (changes requested):
I'm happy with all the changes this pull request currently makes, but the new feature still has to be documented in [`astropy.cosmology` I/O documentation](https://docs.astropy.org/en/latest/cosmology/io.html).

nstarman commented on docs/whatsnew/6.0.rst:
```diff
@@ -65,6 +68,28 @@ method, allowing users to export a cosmology object to a LaTeX table.::
+Renaming fields in I/O
+----------------------
+
+Most I/O methods in :mod:`~astropy.cosmology` (accessed by
```
Done. I added it at the end.

eerovaher reviewed (changes requested):
I have a few small requests regarding the documentation. I think we can merge when they are addressed.

eerovaher reviewed (commented):
Merging the pull request that renamed `astropy.cosmology.io` to `astropy.cosmology._io` has created merge conflicts that need to be solved, but other than that this looks good.

nstarman commented:
Yikes, but failures are unrelated.

pllim commented:
numpy 1.25 is released but h5py didn't catch up. Annoying...

eerovaher commented:
The deprecation warnings caused by `h5py` have been dealt with on `main`, so I suggest rebasing to allow all the CI tests to succeed.

eerovaher commented:
I'd like to merge, but the change log entry CI check is stuck. @pllim, what's the best way to proceed?

pllim commented:
Best way was probably what you did. I merged it. 😉 

Sigh, if only there is a better way to run change log check... it gets stuck more and more often now.

pllim commented:
Thanks, all!

mhvk commented:
I usually just turn labels on and off so it runs again (and nothing else does), but it is indeed a bit of a pain.

pllim commented:
The problem here is we rely on Actions, so it competes with other jobs and sometimes GitHub might flag it as spammy.
